packages/design-system/src/components/NativeDialog/NativeDialog.tsx

```diff
@@ -50,8 +61,18 @@ export const NativeDialog = ({
   useEffect(() => {
     const dialogNode = dialogRef.current;
 
-    // Show the modal
-    showModal ? dialogNode.showModal() : dialogNode.show();
+    // Show or hide the dialog based on `isOpen` value. The `dialogNode.open` property is
+    // a read-only value that will tell us if our dialog DOM element is actually in the
+    // open state.
+    if (isOpen) {
+      if (!dialogNode.open) {
+        showModal ? dialogNode.showModal() : dialogNode.show();
+      }
+    } else {
+      if (dialogNode.open) {
+        dialogNode.close();
+      }
+    }
```

Comment:
Because some of this state comes from the DOM, it didn't lend itself as well to modeling as a finite state machine. I didn't want to duplicate that state and make it way more verbose. Don't @ me. Or do, because this is a review.